Mutual Information I(A,B) and improving doctests/documentation in  `entanglement.jl`

Fe-r-oz opened the pull request:
This PR aims to introduce Mutual Information `I(A,B)` along with improving the doctests and documentation in the `entanglement.jl` .
 
In the future, if you like, it would be better to have a entanglement introduced in the manual of either "algebra" or upcoming "circuit simulation" manual.   

In the papers, there were many graphs, I think based on entanglement entropy that could be reproduced in a tutorial fashion. Hopefully, this sounds interesting.
....
If you want to submit an unfinished piece of work in order to get comments and discuss, please mark the pull request as a draft and ping the repository maintainer.

**Please address only one topic or issue per pull request! Many small PRs are much easier to review and merge than one large PR.**

Before merging, all changes and new functionality should be marked in the CHANGELOG file, but feel free to just leave your CHANGELOG notes in the PR description, to avoid merge conflicts with other requests modifying that file. The maintainer will add these CHANGELOG notes for you if you do so.

Before considering your pull request ready for review and merging make sure that all of the following are completed (please keep the clecklist as part of your PR):

- [x] The code is properly formatted and commented.
- [x] Substantial new functionality is documented within the docs.
- [x] All new functionality is tested.
- [x] All of the automated tests on github pass.


If possible, keep your git history not too wild (rebase and squash commits, keep commits small and semantically separated) so that review is easier.

Krastanov commented on docs/src/references.bib:
```diff
@@ -111,13 +111,13 @@ @article{gullans2020dynamical
-  journal={arXiv preprint quant-ph/0602096},
-  url={https://arxiv.org/abs/quant-ph/0602096},
-  doi={10.48550/arXiv.quant-ph/0602096},
-  year={2006}
```
could you keep the url and doi (so people can find this resource more easily)

Krastanov commented on src/entanglement.jl:
```diff
@@ -132,15 +132,40 @@ end
+
+`B(𝒢) ≡ {(𝓁(g₁),𝓇(g₁)),…,(𝓁(gₙ),𝓇(gₙ))}`
+
+The clipped gauge `𝒢` is a specific choice of stabilizer state where exactly two stabilizer endpoints exist at each site,
+ensuring `ρ𝓁(x) + ρ𝓇(x) = 2` for all sites `x`.
```
I do not think `ρ𝓁(x)` is defined.

Fe-r-oz replied:
Indeed.  Thanks!

`ρ` represents the reduced density matrix for the subsystem under consideration.

Krastanov commented on src/entanglement.jl:
```diff
@@ -162,24 +187,52 @@ end
+Get bipartite entanglement entropy of a subsystem `𝒶`. 
 
-Defined as entropy of the reduced density matrix.
+In a system of `n`-qubits divided into regions `𝒶` and `𝒷`, the Rényi (or von Neumann)
```
We do not use Rényi entropy anywhere here, let's not mention it as it can lead to confusion.

Krastanov commented on src/entanglement.jl:
```diff
@@ -162,24 +187,52 @@ end
 
-Defined as entropy of the reduced density matrix.
+In a system of `n`-qubits divided into regions `𝒶` and `𝒷`, the Rényi (or von Neumann)
+entropy `S(ρ𝒶)` where `ρ𝒶` is reduced density matrix, quantifies the quantum information
+in region `𝒶` after tracing out region `𝒷`.
```
"quantifies the quantum information in region `𝒶` after tracing out region `𝒷`." is not wrong, but it does not really say anything either. You can remove this paragraph and just have a mathematical definition.

Krastanov commented on src/entanglement.jl:
```diff
@@ -162,24 +187,52 @@ end
 
-It can be calculated with multiple different algorithms,
-the most performant one depending on the particular case.
+The entropy is given by: `S(ρ𝒶) = |𝒶| - log₂ |𝒢𝒶|`, where `𝒢𝒶` is a subgroup of stabilizers
+acting non-trivially on `𝒶`. This entropy measures the number of independent stabilizers 
+on `𝒶`,  reflecting the quantum  correlations between regions. 
+
+The stabilizer group `𝒢`, viewed as a binary vector space `V`, can be decomposed such that:
+`S𝒶 = |𝒷| - dim V𝒷 = dim (Π𝒶 V) - |𝒶|` where `Π𝒶` and `Π𝒷` are truncation maps for regions
+`𝒶` and `𝒷`, respectively. 
```
This would be more appropriate in a tutorial or a math lecture. The docstring should probably not include all of this, rather just a quick definition. This is a difficult balance to strike, but my main reason to prefer this to not be included is because it is too difficult to follow (much more difficult than just giving the formula and saying it is also related to the rank of the tableau)

Krastanov commented on src/entanglement.jl:
```diff
@@ -162,24 +187,52 @@ end
+`S𝒶 = |𝒷| - dim V𝒷 = dim (Π𝒶 V) - |𝒶|` where `Π𝒶` and `Π𝒷` are truncation maps for regions
+`𝒶` and `𝒷`, respectively. 
+
+Initially, each site has one stabilizer and zero entanglement, but as stabilizers enter 
+region `𝒶` over time, entanglement `S𝒶` increases by one bit for each new independent
+operator in `𝒶`.
```
similarly getting a bit off topic

Krastanov commented on src/entanglement.jl:
```diff
@@ -227,4 +310,61 @@ function entanglement_entropy(state::AbstractStabilizer, subsystem::AbstractVect
 end
 
+"""
+$TYPEDSIGNATURES
```
Can you post a screenshot of the docs with these additions? I am a bit worried that adding this to every method makes things repetitive (but I am not sure how it renders).

Fe-r-oz replied:
Sorry for that. I will remove all the repetitve `$TYPEDSIGNATURES`. I haven't explored dev docs rendering yet!

Krastanov commented on src/entanglement.jl:
```diff
@@ -227,4 +310,61 @@ function entanglement_entropy(state::AbstractStabilizer, subsystem::AbstractVect
+function mutual_information(state::AbstractStabilizer, A::UnitRange, B::UnitRange, algorithm::Val{:clip}; clip::Bool=true)
+    S𝒶 = entanglement_entropy(state, A, algorithm; clip=clip)
+    S𝒷 = entanglement_entropy(state, B, algorithm; clip=clip) 
+    S𝒶𝒷 = entanglement_entropy(state, _to_unitrange(union(A, B)), algorithm; clip=clip)
```
Why is `_to_unitrange` necessary?

Fe-r-oz replied:
`union `of two `UnitRange` returns a `vector`, while `clip` requires `UnitRange.`

```
julia> typeof(union(1:2, 3:5))
Vector{Int64} (alias for Array{Int64, 1})

```

Fe-r-oz replied:
`_to_unitrange` as helper function is not necessary and it is removed as it was causing one `JET` error when `mutual_information` with `clip` algorithm was called,  which caused two `CI-nightly` tests to fail.

Probably, the function was not a cleaner approach which `JET` pointed out.   Now, it's fixed.

Krastanov commented on src/entanglement.jl:
```diff
@@ -227,4 +310,61 @@ function entanglement_entropy(state::AbstractStabilizer, subsystem::AbstractVect
+
+See Eq. E6 of [li2019measurement](@cite).
+"""
+function mutual_information(state::AbstractStabilizer, A::UnitRange, B::UnitRange, algorithm::Val{:clip}; clip::Bool=true)
```
should probably check that A and B do not overlap

Krastanov commented on src/entanglement.jl:
````diff
@@ -227,4 +310,61 @@ function entanglement_entropy(state::AbstractStabilizer, subsystem::AbstractVect
+    return S𝒶 + S𝒷 - S𝒶𝒷
+end
+
+"""
+$TYPEDSIGNATURES
+
+```jldoctest
+julia> using Graphs # hide
+
+julia> s = Stabilizer(Graph(ghz(4)))
++ XZZZ
++ ZX__
++ Z_X_
++ Z__X
+
+julia> mutual_information(s, [1,2], [3, 4], Val(:graph))
+2
+```
+
+"""
````
Unless the docstrings are specifically to a method, there should probably just be one docstring for the function as a whole.

Fe-r-oz replied:
This is pretty important point. Thanks for letting me know.

Krastanov commented on src/entanglement.jl:
```diff
@@ -227,4 +310,61 @@ function entanglement_entropy(state::AbstractStabilizer, subsystem::AbstractVect
+
+"""
+function mutual_information(state::AbstractStabilizer, A::AbstractVector, B::AbstractVector, algorithm::Union{Val{:rref}, Val{:graph}}; pure::Bool=false)
+    if algorithm == Val(:graph)
```
I think it will probably be cleaner to have separate methods and use dispatch, not if branches

Krastanov reviewed (changes requested):
thanks, this is an interesting addition. I left a few comments in.

Fe-r-oz commented:
Thanks for your comments, I will polish it as per your comments.

Fe-r-oz commented:
All of your suggestions have been incorporated. Thanks for your helpful suggestions.

Krastanov commented on src/QuantumClifford.jl:
```diff
@@ -77,7 +77,7 @@ export
     # Group theory tools
     groupify, minimal_generating_set, pauligroup, normalizer, centralizer, contractor, delete_columns,
     # Clipped Gauge
-    canonicalize_clip!, bigram, entanglement_entropy,
```
let's not export it as this is something we would want to synchronize with QuantumInterface

Could you post an issue at https://github.com/qojulia/QuantumInterface.jl/ describing the need for a new generic function and referencing this PR. The goal would be to pick a name that can be reused by other packages.

In the meantime, keeping this unexported will let us merge this PR without waiting for that coordination.

Fe-r-oz replied:
Thank you! Done!

Krastanov commented on src/entanglement.jl:
```diff
@@ -132,15 +132,41 @@ end
+`B(𝒢) ≡ {(𝓁(g₁),𝓇(g₁)),…,(𝓁(gₙ),𝓇(gₙ))}`
+
+The clipped gauge `𝒢` is a specific choice of stabilizer state where exactly two stabilizer endpoints exist at each site,
+ensuring `ρₗ(x) + ρᵣ(x) = 2` for all sites `x` where `ρ` represents the reduced density matrix for the subsystem under
```
That sentence is unclear to me. How can two density matrices sum up to be the scalar `2`?

Fe-r-oz replied:
Thank you for helpful comment!  I will correct this documentation mistake.

Krastanov commented on src/entanglement.jl:
````diff
@@ -228,3 +283,56 @@ function entanglement_entropy(state::AbstractStabilizer, subsystem::AbstractVect
+```
+
+See Eq. E6 of [li2019measurement](@cite). See also: [`entanglement_entropy`](@ref)
+"""
````
this should probably be a docstring for a function declaration `function mutual_information end`, and not be attached to a specific method. Or probably even better if there is a method with some default choice of algorithm (that way it will not cause issues when this gets declared in QuantumInterface).

Fe-r-oz replied:
Thank you for helpful comments!

Fe-r-oz replied:
Yup, a method with a default choice would be better, so let's use `:clip` as the default algorithm

Krastanov commented on test/test_entanglement.jl:
```diff
@@ -50,4 +50,24 @@
         @test entanglement_entropy(copy(s), subsystem, Val(:rref))==2
     end
+
+    @testset "Mutual information for Clifford circuits" begin
```
additional tests that use QuantumOptics to double check the results independently would be very valuable

Fe-r-oz replied:
Thank you for this very cool suggestion! I will use stuff from `QuantumOpticsBase` to complete this suggestion!

Krastanov reviewed (changes requested):
This looks pretty good.

Apologies for not prioritizing this style of PRs -- you have a lot and I have tried to address the higher value ones first, whenever I have time.

I left a few comments in, mainly motivated by the needs of some other parts of the ecosystem that we try to coordinate with.

Fe-r-oz commented:
Please help review this PR, thank you!  

My apologies for the delay as in adding code review suggestions. I had initially planned to work on the generalized code paper for some time, but I got caught up in them as they seemed quite exciting to read.

Thank you so much for your helpful suggestions! I especially found the suggestion to add `QuantumOpticsBase` tests in  https://github.com/QuantumSavory/QuantumClifford.jl/pull/338#discussion_r1894183567 very helpful and wonderful! I have reduced the code repetition that I did before (my bad) and also added the `QuantumOpticsBase` tests. I hope that the PR is now polished thanks to your helpful codereview suggestions!

There is now warning that `using QOpticsBase.entanglement_entropy` conflicts with`QC.entanglement_entropy ` 😅  in the tests!

Krastanov reviewed (changes requested):
- the addition to docs are great
- the new capability is definitely very useful
- the tests for it are great

Couple of notes to double check:
- the suggested commit simplifies things a bit, but now we need to see what type of error it raises if we try to use `:clip`
- making a default does not need kwargs... You can just do `mutual_information(state::..., rangea::..., rangeb::...) = mutual_information(state, rangea, rangeb, Val{:rref}())`
- why did the test fail to catch the mistake with union of ranges with a large gap between them? Is something missing from the tests


Last thing, let's add `mutual_information` to QuantumInterface.jl and import it here (just make a very small PR to QuantumInterface, bump the version number, add changelog, and it will get merged pretty quickly)

Fe-r-oz commented:
> why did the test fail to catch the mistake with union of ranges with a large gap between them? Is something missing from the tests

Thank you so much for your insights! I’ve fixed an issue where the tests weren’t properly checking for large gaps in the system. The clipped gauge now ensures stabilizers have well-defined left/right endpoints, and the entanglement entropy calculation now correctly errors when non-contiguous regions (e.g., subsystem_rangeA = 1:2, subsystem_rangeB = 5:6) are provided. Additionally, I addressed the out-of-memory errors by using more moderate system sizes, as ket{} becomes computationally expensive for larger systems. The previous implementation was faulty.  My bad 🙏🏼
